process/smartContract/scQueryService.go

```diff
@@ -259,6 +249,31 @@ func (service *SCQueryService) executeScCall(query *process.SCQuery, gasPrice ui
 	return vmOutput, blockInfo, nil
 }
 
+func (service *SCQueryService) recreateTrie(blockRootHash []byte, blockHeader data.HeaderHandler) error {
+	if check.IfNil(blockHeader) {
+		return process.ErrNilBlockHeader
+	}
+
+	accountsAdapter := service.blockChainHook.GetAccountsAdapter()
+	if blockHeader.GetEpoch()+epochDifferenceToConsiderHistory >= service.getCurrentEpoch() {
+		logQueryService.Trace("calling RecreateTrie, for recent history", "block", blockHeader.GetNonce(), "rootHash", blockRootHash)
+		return accountsAdapter.RecreateTrie(blockRootHash)
+	}
+
+	logQueryService.Trace("calling RecreateTrieFromEpoch, for older history", "block", blockHeader.GetNonce(), "rootHash", blockRootHash)
+	holder := holders.NewRootHashHolder(blockRootHash, core.OptionalUint32{Value: blockHeader.GetEpoch(), HasValue: true})
+	return accountsAdapter.RecreateTrieFromEpoch(holder)
```

Comment:
what happens if:
- current block nonce is 1100
- a vm query request for block 1000 is received
- that request is executed in accordance to block's 1000's state
- a new request without block coordinates is received
Will it use the current state or the block 1000's state?
I guess that it won't use the current state, since we recreate the the accountsAdapter's trie, and we also alter the blockchain hook: `service.blockChainHook.SetCurrentHeader(blockHeader)`

Reply:
It will work as we have a critical section whenever we try to execute a VM query. The lines L168-L169 prove that this is happening as the try to acquire a mutex Lock. Without that mutex, the response would almost certainly be wrong.